src/nni_engine.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2019-2022 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+//
+
+#include "nni_engine.hpp"
+
+#include "gp_engine.hpp"
+#include "gp_dag.hpp"
+
+#include "bitset.hpp"
+#include "subsplit_dag.hpp"
+#include "nni_operation.hpp"
+#include "graft_dag.hpp"
+#include "sugar.hpp"
+
+NNIEngine::NNIEngine(GPDAG &dag, GPEngine &gp_engine)
+    : dag_(dag),
+      graft_dag_(std::make_unique<GraftDAG>(dag)),
+      gp_engine_(gp_engine),
+      run_count_(0) {}
+
+// ** Getters
+
+NNIOperation NNIEngine::GetNNIFromAdjacentNNIs(const size_t which_nni) const {
+  auto adjacent_ptr = adjacent_nnis_.begin();
```

Comment:
This is exactly the same code twice with different arguments. A clear case for factoring out into a function. 

Given that this is only used in tests (and isn't really a great way to access) should we just have a set-by-index accessor function in gpdoctest?

Reply:
Factored into gp_doctest lambda.